File: src/server/config/def/PocketMineConfig.php
<?php

namespace pocketcloud\cloud\server\config\def;

use pocketcloud\cloud\config\Config;
use pocketcloud\cloud\config\type\ConfigTypeList;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\config\ServerProperties;
use pocketcloud\cloud\template\TemplateType;
use pocketcloud\cloud\util\Utils;

final class PocketMineConfig implements ServerProperties {

    public function modify(string $filePath, array $updatedContent): bool {
        $config = new Config($filePath, ConfigTypeList::YML());
        foreach ($updatedContent as $name => $value) {
            $config->set($name, $value);
        }

        return $config->save();
    }

    public function renew(string $filePath): bool {
        $config = new Config($filePath, ConfigTypeList::YML());
        $content = $config->getAll();
        Utils::fillMissingKeys($content, $this->getDefaultContent());
        $config->setAll($content);
        return $config->save();
    }

    public function needsRenewal(string $filePath): bool {
        if (file_exists($filePath)) {
            $config = new Config($filePath, ConfigTypeList::YML());
            $defaultContent = $this->getDefaultContent();
            $currentContent = $config->getAll();
            return !Utils::hasAllKeys($currentContent, $defaultContent);
        }

        return true;
    }

    public function replacePlaceholders(CloudServer $server): array {
        return [];
    }

    public function getDefaultContent(): array {
        return [
            "settings" => [
                "force-language" => false,
                "shutdown-message" => "Server closed",
                "query-plugins" => true,
                "enable-profiling" => false,
                "profile-report-trigger" => 20,
                "async-workers" => "auto",
                "enable-dev-builds" => false
            ],
            "memory" => [
                "global-limit" => 0,
                "main-limit" => 0,
                "main-hard-limit" => 1024,
                "async-worker-hard-limit" => 256,
                "check-rate" => 20,
                "continuous-trigger" => true,
                "continuous-trigger-rate" => 30,
                "garbage-collection" => [
                    "period" => 36000
                ],
                "memory-dump" => [
                    "dump-async-worker" => true
                ],
                "max-chunks" => [
                    "chunk-radius" => 4
                ],
            ],
            "network" => [
                "batch-threshold" => 256,
                "compression-level" => 6,
                "async-compression" => false,
                "async-compression-threshold" => 10000,
                "upnp-forwarding" => false,
                "max-mtu-size" => 1492,
                "enable-encryption" => true
            ],
            "debug" => [
                "level" => 1
            ],
            "player" => [
                "save-player-data" => true,
                "verify-xuid" => true
            ],
            "level-settings" => [
                "default-format" => "leveldb"
            ],
            "chunk-sending" => [
                "per-tick" => 4,
                "spawn-radius" => 4
            ],
            "chunk-ticking" => [
                "tick-radius" => 3,
                "blocks-per-subchunk-per-tick" => 3,
                "disable-block-ticking" => []
            ],
            "chunk-generation" => [
                "population-queue-size" => 32
            ],
            "ticks-per" => [
                "autosave" => 6000
            ],
            "auto-report" => [
                "enabled" => true,
                "send-code" => true,
                "send-settings" => true,
                "send-phpinfo" => false,
                "use-https" => true,
                "host" => "crash.pmmp.io"
            ],
            "anonymous-statistics" => [
                "enabled" => false,
                "host" => "stats.pocketmine.net"
            ],
            "auto-updater" => [
                "enabled" => true,
                "on-update" => [
                    "warn-console" => true
                ],
                "preferred-channel" => "stable",
                "suggest-channels" => true,
                "host" => "update.pmmp.io"
            ],
            "timings" => [
                "host" => "timings.pmmp.io"
            ],
            "console" => [
                "enable-input" => true,
                "title-tick" => true
            ],
            "aliases" => [],
            "worlds" => [],
            "plugins" => [
                "legacy-data-dir" => false
            ]
        ];
    }

    public function getFileName(): string {
        return "pocketmine.yml";
    }

    public function getTemplateType(): TemplateType {
        return TemplateType::SERVER();
    }
}

File: src/main/java/de/pocketcloud/cloud/server/config/impl/PocketMineConfig.java
package de.pocketcloud.cloud.server.config.impl;

import de.pocketcloud.cloud.server.CloudServer;
import de.pocketcloud.cloud.server.config.ServerProperties;
import de.pocketcloud.cloud.server.software.ServerSoftware;
import de.pocketcloud.cloud.server.software.ServerSoftwareManager;
import de.pocketcloud.cloud.util.ArrayUtils;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class PocketMineConfig extends ServerProperties {

    @Override
    public boolean modify(String filePath, Map<String, Object> updatedContent) {
        return modifyYaml(filePath, updatedContent);
    }

    @Override
    public boolean renew(String filePath) {
        return renewYaml(filePath);
    }

    @Override
    public boolean needsRenewal(String filePath) {
        return needsRenewalYaml(filePath);
    }

    @Override
    public Map<String, Object> replacePlaceholders(CloudServer server) {
        return Map.of();
    }

    @Override
    public Map<String, Object> getDefaultContent() {
        return ArrayUtils.orderedMap(
                "settings", ArrayUtils.orderedMap(
                        "force-language", false,
                        "shutdown-message", "Server closed",
                        "query-plugins", true,
                        "enable-profiling", false,
                        "profile-report-trigger", 20,
                        "async-workers", "auto",
                        "enable-dev-builds", false
                ),
                "memory", ArrayUtils.orderedMap(
                        "global-limit", 0,
                        "main-limit", 0,
                        "main-hard-limit", 1024,
                        "async-worker-hard-limit", 256,
                        "check-rate", 20,
                        "continuous-trigger", true,
                        "continuous-trigger-rate", 30,
                        "garbage-collection", ArrayUtils.orderedMap("period", 36000),
                        "memory-dump", ArrayUtils.orderedMap("dump-async-worker", true),
                        "max-chunks", ArrayUtils.orderedMap("chunk-radius", 4)
                ),
                "network", ArrayUtils.orderedMap(
                        "batch-threshold", 256,
                        "compression-level", 6,
                        "async-compression", false,
                        "async-compression-threshold", 10000,
                        "upnp-forwarding", false,
                        "max-mtu-size", 1492,
                        "enable-encryption", true
                ),
                "debug", ArrayUtils.orderedMap("level", 1),
                "player", ArrayUtils.orderedMap("save-player-data", true, "verify-xuid", true),
                "level-settings", ArrayUtils.orderedMap("default-format", "leveldb"),
                "chunk-sending", ArrayUtils.orderedMap("per-tick", 4, "spawn-radius", 4),
                "chunk-ticking", ArrayUtils.orderedMap(
                        "tick-radius", 3,
                        "blocks-per-subchunk-per-tick", 3,
                        "disable-block-ticking", List.of()
                ),
                "chunk-generation", ArrayUtils.orderedMap("population-queue-size", 32),
                "ticks-per", ArrayUtils.orderedMap("autosave", 6000),
                "auto-report", ArrayUtils.orderedMap(
                        "enabled", true,
                        "send-code", true,
                        "send-settings", true,
                        "send-phpinfo", false,
                        "use-https", true,
                        "host", "crash.pmmp.io"
                ),
                "anonymous-statistics", ArrayUtils.orderedMap("enabled", false, "host", "stats.pocketmine.net"),
                "auto-updater", ArrayUtils.orderedMap(
                        "enabled", true,
                        "on-update", ArrayUtils.orderedMap("warn-console", true),
                        "preferred-channel", "stable",
                        "suggest-channels", true,
                        "host", "update.pmmp.io"
                ),
                "timings", ArrayUtils.orderedMap("host", "timings.pmmp.io"),
                "console", ArrayUtils.orderedMap("enable-input", true, "title-tick", true),
                "aliases", new LinkedHashMap<>(),
                "worlds", new LinkedHashMap<>(),
                "plugins", ArrayUtils.orderedMap("legacy-data-dir", false)
        );
    }

    @Override
    public String getFileName() {
        return "pocketmine.yml";
    }

    @Override
    public ServerSoftware getServerSoftware() {
        return ServerSoftwareManager.instance().get("pmmp-latest");
    }
}